src/rust/kafka-metrics-exporter/src/lib.rs

```diff
@@ -0,0 +1,265 @@
+                        Ok((partition, offset)) => {
+                            tracing::debug!(
+                                message="Metric published",
+                                topic=%&topic,
+                                partition=?partition,
+                                offset=?offset,
+                            );
+                        }
+                        Err((e, _)) => {
+                            tracing::error!(
+                                message="Failed to send message to kafka",
+                                topic=%&topic,
+                                error=?e,
+                            );
+                        }
+                    }
+                }
+                producer.flush(Duration::from_secs(3));
+            }
+            .instrument(tracing::debug_span!("recv loop")),
+        );
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug)]
+struct Metric {
+    payload: MetricWrapper,
+    timestamp: i64,
+    key: u64,
+}
+
+#[derive(Clone)]
+struct MetricsBuffer {
+    buffer: UnboundedSender<Metric>,
+}
+
+impl MetricsBuffer {
+    fn new(buffer: UnboundedSender<Metric>) -> Self {
+        Self { buffer }
+    }
+
+    fn push(&self, metric: Metric) {
+        if let Err(SendError(_)) = self.buffer.send(metric) {
+            tracing::warn!(
+                message = "Failed to send message to Kafka, MetricsBuffer receiver was dropped",
+            );
+        };
+    }
+}
+
+#[derive(Clone)]
+struct KafkaRecorder {
+    buffered_metrics: MetricsBuffer,
+    clock: quanta::Clock,
+}
+
+impl KafkaRecorder {
+    fn new() -> (Self, UnboundedReceiver<Metric>) {
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+
+        (
+            Self {
+                buffered_metrics: MetricsBuffer::new(tx),
+                clock: quanta::Clock::new(),
+            },
+            rx,
+        )
+    }
+
+    fn make_timestamp(&self) -> i64 {
+        // Note that this timestamp may be slightly behind reality
+        self.clock.recent().as_unix_duration().as_millis() as i64
+    }
+}
+
+impl Recorder for KafkaRecorder {
+    fn register_counter(&self, key: &Key, unit: Option<Unit>, description: Option<&'static str>) {
+        tracing::trace!(
+            message="register_counter - does nothing for KafkaMetricExporter",
+            name=&key.name(),
+            unit=?unit,
+            description=?description,
+        );
+    }
+
+    fn register_gauge(&self, key: &Key, unit: Option<Unit>, description: Option<&'static str>) {
+        tracing::trace!(
+            message="register_gauge - does nothing for KafkaMetricExporter",
+            name=&key.name(),
+            unit=?unit,
+            description=?description,
+        );
+    }
+
+    fn register_histogram(&self, key: &Key, unit: Option<Unit>, description: Option<&'static str>) {
+        tracing::trace!(
+            message="register_histogram - does nothing for KafkaMetricExporter",
```

Comment:
Why can we not register things, but can update them?

Reply:
Registering is just where you set descriptions for metrics, for systems like prometheus I think this is more important. For this type of metric there's nothing we'd want to register - any registration is going to have to happen out of band.